### DISCUSSION TITLE
Comments from CMS members on HS3 v0.2

### DISCUSSION THREAD

**Discussion by anigamova** [hep-statistics-standard/hep-statistics-standard#35] [Standard]
We have reviewed the latest draft v0.2 within the CMS Collaboration, please find the comments below. 
Also available in [gdoc](https://docs.google.com/document/d/1xSBqNjYmTEtV4s7Lo-C58kx10LTAwAUf5hRuPU4BVec/edit?usp=sharing) in case one find it easier to comment there. 

Contributors:
Nick Smith, Kirill Skovpen, Aliya Nigamova, Kyle Cormier, Clemens Lange, Piergiulio Lenzi, Nick Wardle

### General: 
It would be helpful to include line numbers in the pdf to make commenting easier in future versions
It is not clear if this format is suitable for large models, perhaps this document should provide some recommendations in case of issues.  

### Introduction

- This would be a good place to cite https://arxiv.org/abs/2109.04981 and also repeat some of the justification presented there. Leave the details of software (pyhf and RooFit in 6.26) to a second paragraph. “Sorely” is a bit harsh.

- Can you declare what tool versions conform to the current schema? Probably it is not sufficiently stable yet, but in the future you might advertise here what tools have adopted it.

- Can you introduce notation explicitly for the variates/observables and observable space (e.g. x ~ m(\theta), x \in X)? Later in section 1.1.4 you finally introduce a PDF but I don’t really understand why you would not identify m(\theta) as p(x; \theta) (or p(x|\theta) in Bayesian analyses) here already?
- In the discussion of the structure of \theta, named tuples for now sounds great. But we may want to have a canonical representation of the gradient and hessian, for the purpose of serializing the local structure of best-fit or other significant points in the parameter space. So part of the standard might be a recipe for structured Hessian arrays when they arise. A good example implementation of this is jax pytrees.
- We already have cases in CMS where we need to minimize over a parameter space that includes discrete-valued dimensions (i.e. discrete profiling https://arxiv.org/pdf/1408.6865.pdf), and may someday need to minimize over a manifold. Some thought should go into how the parameter space is defined. Right now, it appears to be only a rectangular subset of R^n
- I think we should elevate m(\theta) to be a proper probability distribution in the observable space, i.e. define the appropriate measure (a differential if density, or counting measure for discrete outcomes) that corresponds to the observable space and cleanly separate what are parameters and what are random variates. In Bayesian analysis, some parameters are promoted to random variates, so I suppose in the analyses section we have to declare which are which.
- I appreciate the measure-theoretic approach in section 1.1.2, but I fear it comes out of order. Perhaps it would be better to start with this and work towards the definition of m(\theta)?

### Section 2.1

- Sec 2.1: The text and equations concerning normalisation are confusing, in particular:
- Why for some distributions the normalisation factor M is given explicitly whereas others it is left out. Additionally, does it not always depend on the domain?
- It is unclear what is meant by saying the distributions MUST be normalised. Normalised to what and when? The normalisation depends on the domain, which as I understand can be dynamically chosen (e.g. in the analysis section) so e.g. it seems normalising them in the definition is not possible.
- Is the distribution name unique among the collection of distributions or all named objects?
- There should be a discussion of the mathematics behind extended distributions in the introduction, with a reference to https://doi.org/10.1016/0168-9002(90)91334-8 Personally, I would advocate introducing unbinned data as a Poisson process, and point out that the extended likelihood is just the [Poisson point process](https://en.wikipedia.org/wiki/Poisson_point_process) rate function. Perhaps we can define distributions and Poisson processes separately, rather than allowing some distributions to not be distributions but these extended objects that are unique to HEP? That would make it much clearer when we are treating data as unbinned and when we are treating it binned. This would also motivate having the domain of the observable space (i.e the axes min and max in unbinned data, section 2.3.2) be part of the Poisson process definition rather than data, since after all that does not change with repeated experiments.
- 2.1.1: Are x and (a,b) mutually exclusive arguments? Otherwise there is a potential for mismatch between the domain of x and the (a,b) values
- 2.1.3: a non-integer Poisson distribution is out of place here, or at least I’ve never heard of it being used to describe the expected distribution of physical measurements. Rather this is a topic that comes up for Asimov datasets, which should also be discussed in the introduction. Really, Asimov datasets are an approximation tool for some test statistic distributions.
- 2.1.6: why is “covariances” plural? I think it should be singular
- 2.1.7: it would be good to eventually extend this to other bases, such as Chebyshev or Bernstein
- 2.1.9: this seems out of place, it is something to do with binned fits. Anyway, isn’t this just a vectorized Gamma distribution? Anyway, BB-lite can be implemented also using Normal distributions if the effective entries is sufficiently high, so it shouldn’t be pigeonholed here
- 2.1.10: introduce h(z) after they are first used!
- 2.1.11: the case where the distribution is not extended is an excellent example where manifold optimization is helpful, because the correlation introduced by Eqn. 14 can really cause trouble for numeric minimization
- I would separate 2.1.11 and 2.1.12 from the rest as higher-order distributions. Along with sum and product, I expected to find a convolution. Another convenient higher-order distribution is one that takes a function and defines the subset of parameters to promote to random variates (i.e. observables), creating either a distribution or a Poisson process rate (if extended).
- I would group all exponential-family distributions together
- Sec 2.1.14 It is unclear why common functions (exp, pow, min, max) should not have defined behaviour and only MAY be supported. 
- Sec 2.1.14 “SHOULD not” should be written as “SHOULD NOT”
- Sec 2.1.14 What is the meaning of “MAY only” in the context of a standard, does it mean MUST?
- Sec 2.1.15 It is unclear how one correlates systematic uncertainties, e.g. if two modifiers for different samples or in different channels should share a common constraint term. Is it by name? And if so, what happens if the same constraint name is given, but its functional form is specified differently? Is there another mechanism?
- Sec 2.1.15 It would be good if the standard could allow a richer variety of correlations for systematic uncertainties 
- Sec 2.1.15 How is the interpolation between different values of HistFactory modifiers such as histosys to be performed?
- 2.1.15 Why do we have Eqn. 21 when this is introduced with respect to a single bin at the end of the opening paragraph?
- 2.1.15 Is there no Gamma constraint term?
- 2.1.15 What about axes with irregular binning? I think the specification here should match exactly that of 2.3.3
- 2.1.15 Allow for alternative stat errors rather than only Barlow-Beeston-lite 

### Section 2.3

- You need to add in the introduction what the statistical interpretation of these “uncertainty” values are. In most settings, data has no uncertainty. The uncertainty in our measurement of some value is encoded in the likelihood model (e.g. via a convolution with a Gaussian)
- Is there not a duplication of information between what may appear in this section and section 2.5 regarding the domain of the data? Since an observation and its corresponding PDF have to share the domain (or better, manifold) it would be nice to define it once.

### Section 2.4

- The usage of “likelihood” in this context is not statistically accurate, perhaps better to call this section “statistical model”
- Perhaps a better term could be “aux_constraints” instead of “aux_distributions” since they aren’t really distributions (what is the random variate?) One might find here things like Lagrange multipliers or regularization terms. In general there are soft and hard constraints, i.e. penalty terms or equations that have to remain satisfied. This might also be a good place to put manifold definitions, though those (hard) constraints might also be implicitly defined by section 2.5.

### Section 2.5

- These sound like manifolds. Can we call them that?

**Comment by cburgard** [comment]
Dear Nick S., Kirill, Aliya, Kyle, Clemens, Piergiulio, Nick W.,

Thank you very much for your extremely detailed and thoughtful comments. We have taken some time to reflect on them, implement your suggestions as far as possible and formulate replies.

We would specifically like to thank you for your effort in improving the statistical rigor, clarity of language and fundamental approach we have taken with this document. We feel that your contributions have been extremely valuable, and we hope that you are satisfied with our responses and the corresponding changes in the document.

A new version of the document has also been attached to this message. Please feel free to post further comments in the future, either if you disagree with any of our replies, want to suggest an amendment, or want to comment on new points not previously addressed or other new material added to the document.

Thank you very much for your kind review!

Regards,
Carsten Burgard (with the help of Oliver Schulz @oschulz and Cornelius Grunewald @Cornelius-G )
[HS3.pdf](https://github.com/hep-statistics-serialization-standard/hep-statistics-serialization-standard/files/15063517/HS3.pdf)

-------------------------------

### **General:**

It would be helpful to include line numbers in the pdf to make commenting easier in future versions

→done \
 It is not clear if this format is suitable for large models, perhaps this document should provide some recommendations in case of issues.

→the introductory text has been slightly adjusted to make the point that large models are within the scope of this standard


### **Introduction**


* This would be a good place to cite[ https://arxiv.org/abs/2109.04981](https://arxiv.org/abs/2109.04981) and also repeat some of the justification presented there. Leave the details of software (pyhf and RooFit in 6.26) to a second paragraph. “Sorely” is a bit harsh. \
→wording has been adjusted \

* Can you declare what tool versions conform to the current schema? Probably it is not sufficiently stable yet, but in the future you might advertise here what tools have adopted it. \
→unfortunately, this is hard to do, since the document will always need to be finalized before the tools are (in each version). however, we added a sentence inviting the tool developers to cite versions of the standard. \

* Can you introduce notation explicitly for the variates/observables and observable space (e.g. x ~ m(\theta), x \in X)?  \
→ done (first paragraph of 1.2.1, is this what you had in mind?)
* Later in section 1.1.4 you finally introduce a PDF but I don’t really understand why you would not identify m(\theta) as p(x; \theta) (or p(x|\theta) in Bayesian analyses) here already? \
→clarified PDF notation in last paragraph of 1.2.2. we avoid the notation p(x|theta) as it does not include “m” and does not clearly distinguish between finite and differential probabilities \

* In the discussion of the structure of \theta, named tuples for now sounds great. But we may want to have a canonical representation of the gradient and hessian, for the purpose of serializing the local structure of best-fit or other significant points in the parameter space. So part of the standard might be a recipe for structured Hessian arrays when they arise. A good example implementation of this is jax pytrees. \
→we agree that gradients and hessians will become relevant. we expect that such calculations will however typically take place in a transformed space, where variable names and (more importantly) variable bounds have been transformed away, since gradient-based algorithms are typically not compatible with bound spaces \

* We already have cases in CMS where we need to minimize over a parameter space that includes discrete-valued dimensions (i.e. discrete profiling[ https://arxiv.org/pdf/1408.6865.pdf](https://arxiv.org/pdf/1408.6865.pdf)), and may someday need to minimize over a manifold. Some thought should go into how the parameter space is defined. Right now, it appears to be only a rectangular subset of R^n \
→we expect that in the Bayesian case, this will be handled by using discrete distributions for such parameters, which the standard itself already allows (though implementations would have to support it). In the frequentist case, we would need to add the ability to specify sets of allowed parameter values (right now, only upper and lower bounds are supported). It will be easy to extend the standard in the future to include this. Please feel free to make syntax suggestions. \

* I think we should elevate m(\theta) to be a proper probability distribution in the observable space, i.e. define the appropriate measure (a differential if density, or counting measure for discrete outcomes) that corresponds to the observable space and cleanly separate what are parameters and what are random variates. In Bayesian analysis, some parameters are promoted to random variates, so I suppose in the analyses section we have to declare which are which. \
→We do indeed understand m(theta) to always be a proper probability distribution / measure, which comes with a standard PDF in respect to an implied reference measure. By default, discrete distributions are referring to the counting measure, continuous ones are referring to Lebesque. Should we state this implied reference for each distribution defined by the standard? As for the Bayesian case, we do not think that distributions for parameters and data need to be expressed in different ways. All distributions are specified in the same way, whether they describe parameters or data is specified at the “point of use”. \

* I appreciate the measure-theoretic approach in section 1.1.2, but I fear it comes out of order. Perhaps it would be better to start with this and work towards the definition of m(\theta)? \
→As measure theory is (sadly) often not taught as part of a physicist’s standard statistics curriculum, we feel that putting it front and center might make the document less accessible. Personally, we would not mind starting with measure theory - but to make the standard appealing to as wide an audience as possible, we think it might be better to work our way up to it. This way, people not familiar with measure theory can skip over the corresponding paragraphs in the standard. \



### **Section 2.1**

* Why for some distributions the normalisation factor M is given explicitly whereas others it is left out. Additionally, does it not always depend on the domain? \
→removed everywhere, explanation of the domains extended
* It is unclear what is meant by saying the distributions MUST be normalised. Normalised to what and when? The normalisation depends on the domain, which as I understand can be dynamically chosen (e.g. in the analysis section) so e.g. it seems normalising them in the definition is not possible. \
→correct, we fixed this, see above.
* Is the distribution name unique among the collection of distributions or all named objects? \
→yes, we have tried to improve the clarity of the text
* There should be a discussion of the mathematics behind extended distributions in the introduction, with a reference to[ https://doi.org/10.1016/0168-9002(90)91334-8](https://doi.org/10.1016/0168-9002(90)91334-8) Personally, I would advocate introducing unbinned data as a Poisson process, and point out that the extended likelihood is just the[ Poisson point process](https://en.wikipedia.org/wiki/Poisson_point_process) rate function. Perhaps we can define distributions and Poisson processes separately, rather than allowing some distributions to not be distributions but these extended objects that are unique to HEP? That would make it much clearer when we are treating data as unbinned and when we are treating it binned. This would also motivate having the domain of the observable space (i.e the axes min and max in unbinned data, section 2.3.2) be part of the Poisson process definition rather than data, since after all that does not change with repeated experiments. \
→ we agree that primary terminology should be “Poisson point process”. We have rewritten the corresponding part of the standard completely to reflect this preference. We have added a reference to the paper you suggested. 
* 2.1.1: Are x and (a,b) mutually exclusive arguments? Otherwise there is a potential for mismatch between the domain of x and the (a,b) values \
→they should never have been included in the first place and are now removed
* 2.1.3: a non-integer Poisson distribution is out of place here, or at least I’ve never heard of it being used to describe the expected distribution of physical measurements. Rather this is a topic that comes up for Asimov datasets, which should also be discussed in the introduction. Really, Asimov datasets are an approximation tool for some test statistic distributions. \
→we have split the distribution into two now
* 2.1.6: why is “covariances” plural? I think it should be singular \
→it’s supposed to relate to the entries of the covariance matrix. the wording has been improved to make this clear.
* 2.1.7: it would be good to eventually extend this to other bases, such as Chebyshev or Bernstein \
→that’s correct, other types of polynomial distributions (e.g. chebychev_dist and bernstein_dist) will be added in the future
* 2.1.9: this seems out of place, it is something to do with binned fits. Anyway, isn’t this just a vectorized Gamma distribution? Anyway, BB-lite can be implemented also using Normal distributions if the effective entries is sufficiently high, so it shouldn’t be pigeonholed here \
→for many implementations, it’s very advantageous to have a pre-made type for this type of distribution. \
adding a variant that instead uses a gaussian (e.g. beeston_barlow_gauss_constraint_dist) might be useful \
however, what exactly we will do with this type of distribution also depends on the outcome of the discussion about global observables, so we will postpone any changes to this until that discussion is resolved: [https://github.com/hep-statistics-serialization-standard/hep-statistics-serialization-standard/discussions/20](https://github.com/hep-statistics-serialization-standard/hep-statistics-serialization-standard/discussions/20) \
For now, we have renamed this into “beeston-barlow_poisson_constraint_dist to honor the distinction.
* 2.1.10: introduce h(z) after they are first used! \
→ moved
* 2.1.11: the case where the distribution is not extended is an excellent example where manifold optimization is helpful, because the correlation introduced by Eqn. 14 can really cause trouble for numeric minimization \
→ we now explicitly allow for a number of coefficients equal to the number of distributions to be passed also in the non-extended case. implementations will then have to use methods like manifold optimization or, in the Bayesian case, users will need to specify priors like Dirichlet distributions (still to be added) that guarantee that the coefficients sum up to one. we keep the default behavior for the number of coefficients smaller than the number of distributions for compatibility with ROOT (and similar), though indeed this approach is not ideal
* I would separate 2.1.11 and 2.1.12 from the rest as higher-order distributions. Along with sum and product, I expected to find a convolution. Another convenient higher-order distribution is one that takes a function and defines the subset of parameters to promote to random variates (i.e. observables), creating either a distribution or a Poisson process rate (if extended). \
→ we have opted to group the distributions into univariate fundamental, multivariate fundamental and composite distributions
* I would group all exponential-family distributions together \
→ see above. we feel that everything involving an exponential function is not a very useful category, as this would also involve e.g. the crystal ball function (which includes an exponential tail)
* Sec 2.1.14 It is unclear why common functions (exp, pow, min, max) should not have defined behaviour and only MAY be supported. \
→ we now tried to clarify the role of generic distributions. the behavior of all listed functions is now clearly defined, unified with the generic function, and externalized into a new section. the list of supported functions is open to extension in further versions of the standard
* Sec 2.1.14 “SHOULD not” should be written as “SHOULD NOT” \
→fixed
* Sec 2.1.14 What is the meaning of “MAY only” in the context of a standard, does it mean MUST? \
→fixed
* Sec 2.1.15 It is unclear how one correlates systematic uncertainties, e.g. if two modifiers for different samples or in different channels should share a common constraint term. Is it by name? And if so, what happens if the same constraint name is given, but its functional form is specified differently? Is there another mechanism? \
→a sentence has been added to clarify that correlations are induced by parameter_name, and that they have to share the same constraint term in such cases. Future versions of the standard will likely move the definition of the constraint term to a different location to avoid duplication and ambiguity.
* Sec 2.1.15 It would be good if the standard could allow a richer variety of correlations for systematic uncertainties \
→this is an interesting suggestion which has been made many times also for the ROOT implementation of HistFactory, and which has generally resulted in intense discussions. we would like to invite a specific discussion on this topic to invite input reflecting as many different points of view as possible.
* Sec 2.1.15 How is the interpolation between different values of HistFactory modifiers such as histosys to be performed? \
→ a segment on interpolation functions has been added
* 2.1.15 Why do we have Eqn. 21 when this is introduced with respect to a single bin at the end of the opening paragraph? \
→ thanks for catching this, the text now related to the entire histogram instead
* 2.1.15 Is there no Gamma constraint term? \
→ yes there is, it is handled alongside all other constraint terms
* 2.1.15 What about axes with irregular binning? I think the specification here should match exactly that of 2.3.3 \
→ thanks for the suggestion, the change has been done as requested
* 2.1.15 Allow for alternative stat errors rather than only Barlow-Beeston-lite \
→ this is a good suggestion which will be adopted in future versions of the standard. please suggest extensions as issues or open a discussion on this point.


### **Section 2.3**



* You need to add in the introduction what the statistical interpretation of these “uncertainty” values are. In most settings, data has no uncertainty. The uncertainty in our measurement of some value is encoded in the likelihood model (e.g. via a convolution with a Gaussian) \
→ A paragraph to this extent has been added (lines 596-605)
* Is there not a duplication of information between what may appear in this section and section 2.5 regarding the domain of the data? Since an observation and its corresponding PDF have to share the domain (or better, manifold) it would be nice to define it once. (TODO remove ranges for unbinned data) \
→ the duplication here is unavoidable and might even be beneficial to ensure that the data matches the model regarding the binning. the model and the data may often stem from different sources, so duplication of information is useful to enforce cross-checks, which should be mandatory for implementations.


### **Section 2.4**



* The usage of “likelihood” in this context is not statistically accurate, perhaps better to call this section “statistical model” \
→we would like to keep the term likelihood here. in the Bayesian case, e.g., the statistical model would also include the priors, which are combined with the likelihood in the “analysis” component. combining parametrized distributions m(theta) with observed data x that corresponds to them does generate likelihood functions PDF(m(theta),x) with a constant x and varying theta, so we think the terminology is correct. we have clarified this in the standard text now. please check if it finds your approval.
* Perhaps a better term could be “aux_constraints” instead of “aux_distributions” since they aren’t really distributions (what is the random variate?) One might find here things like Lagrange multipliers or regularization terms. In general there are soft and hard constraints, i.e. penalty terms or equations that have to remain satisfied. This might also be a good place to put manifold definitions, though those (hard) constraints might also be implicitly defined by section 2.5. \
→ we have clarified now that the random variate for the aux_distributions does exist, but is implicit. this is, however, an interesting point still under discussion. please participate in the corresponding discussion on our github!


### **Section 2.5**



* These sound like manifolds. Can we call them that? \
→ we would understand the term “manifolds” more as referring to the properties of sets of values, while the term “domain” we think more clearly indicates how we use them, i.e., as sets of allowed parameter values. the term “domain” may also be more accessible to the average user (who may think of these as parameter ranges mostly)
